diff --git a/src/3d/chunks/qgschunkedentity_p.cpp b/src/3d/chunks/qgschunkedentity_p.cpp
index fe78f64cd4b..16b61abad8b 100644
--- a/src/3d/chunks/qgschunkedentity_p.cpp
+++ b/src/3d/chunks/qgschunkedentity_p.cpp
@@ -30,6 +30,137 @@
 
 ///@cond PRIVATE
 
+static QVariantList dumpBbox( const QgsAABB &bbox )
+{
+  QVariantList vl;
+  vl << bbox.xMin << bbox.yMin << bbox.zMin << bbox.xMax << bbox.yMax << bbox.zMax;
+  return vl;
+}
+
+struct ChunkNodeInfo
+{
+  QgsChunkNodeId nodeId;
+  float nodeError;
+  QgsAABB nodeBbox;
+  QgsChunkNode::State nodeState;
+
+  QVariantMap dump() const
+  {
+    QVariantMap vm;
+    vm["id"] = nodeId.text();
+    vm["error"] = nodeError;
+    vm["bbox"] = dumpBbox( nodeBbox );
+    vm["state"] = ( int ) nodeState;
+    return vm;
+  }
+};
+
+struct EntityUpdateRecord
+{
+  QTime timeUpdateStart, timeUpdateFinish;
+  Qgs3DMapSceneEntity::SceneContext sceneContext;
+  // state at the end
+  QVector<QgsChunkNodeId> activeNodes;
+  QVector<ChunkNodeInfo> replacementQueueNodes;   // in the order in the replacement queue
+  QVector<ChunkNodeInfo> loadingQueueNode;        // in the order in the loading queue
+  QVector<ChunkNodeInfo> activeJobsNodes;      // nodes as active jobs (now being loaded)
+
+  // TODO? pruned nodes at start, unloaded nodes at the end
+};
+
+#include <nlohmann/json.hpp>
+#include "qgsjsonutils.h"
+
+struct EntityLog
+{
+  QTime timeLogStart;
+  QVector<EntityUpdateRecord> records;
+
+  EntityLog()
+  {
+    timeLogStart = QTime::currentTime();
+  }
+
+  EntityUpdateRecord &currentRecord() { return records.last(); }
+
+  static QVariantList vector2variant( const QVector3D &v )
+  {
+    QVariantList lst;
+    lst << v.x() << v.y() << v.z();
+    return lst;
+  }
+
+  void toJson( const QString &outputFilename )
+  {
+    QVariantList recs;
+    for ( const EntityUpdateRecord &r : std::as_const( records ) )
+    {
+      QVariantMap r2;
+      r2["timeStart"] = r.timeUpdateStart.toString( "hh:mm:ss.zzz" );
+      r2["timeFinish"] = r.timeUpdateFinish.toString( "hh:mm:ss.zzz" );
+      r2["camera"] = vector2variant( r.sceneContext.cameraPos );
+      // TODO: camera matrix
+
+      QVariantList activeLst;
+      for ( const QgsChunkNodeId &nid : std::as_const( r.activeNodes ) )
+        activeLst << nid.text();
+      r2["active"] = activeLst;
+
+      QVariantList loadLst;
+      for ( const ChunkNodeInfo &ni : std::as_const( r.loadingQueueNode ) )
+        loadLst << ni.dump();
+      r2["loading"] = loadLst;
+
+      QVariantList replLst;
+      for ( const ChunkNodeInfo &ni : std::as_const( r.replacementQueueNodes ) )
+        replLst << ni.dump();
+      r2["replacement"] = replLst;
+
+      QVariantList jobsLst;
+      for ( const ChunkNodeInfo &ni : std::as_const( r.activeJobsNodes ) )
+        jobsLst << ni.dump();
+      r2["jobs"] = jobsLst;
+
+      recs.append( r2 );
+    }
+
+    QVariantMap var;
+    var["rec"] = recs;
+    var["timeStart"] = timeLogStart.toString( "hh:mm:ss.zzz" );
+
+    json j = QgsJsonUtils::jsonFromVariant( var );
+    QFile f( outputFilename );
+    if ( !f.open( QIODevice::WriteOnly ) )
+    {
+      Q_ASSERT( false );
+      return;  // TODO
+    }
+    f.write( QByteArray::fromStdString( j.dump() ) );
+  }
+};
+
+ChunkNodeInfo _dumpChunkInfo( QgsChunkNode *ch )
+{
+  ChunkNodeInfo n;
+  n.nodeId = ch->tileId();
+  n.nodeBbox = ch->bbox();
+  n.nodeError = ch->error();
+  n.nodeState = ch->state();
+  return n;
+}
+
+void _dumpList( QgsChunkList *lst, QVector<ChunkNodeInfo> &output )
+{
+  QgsChunkListEntry *ptr = lst->first();
+  while ( ptr )
+  {
+    QgsChunkNode *ch = ptr->chunk;
+    output.append( _dumpChunkInfo( ch ) );
+    ptr = ptr->next;
+  }
+}
+
+
 
 static float screenSpaceError( QgsChunkNode *node, const QgsChunkedEntity::SceneContext &sceneContext )
 {
@@ -70,6 +201,8 @@ QgsChunkedEntity::QgsChunkedEntity( Qgs3DMapSettings *mapSettings, float tau, Qg
   mChunkLoaderQueue = new QgsChunkList;
   mReplacementQueue = new QgsChunkList;
 
+  mLog.reset( new EntityLog );
+
   // in case the chunk loader factory supports fetching of hierarchy in background (to avoid GUI freezes)
   connect( loaderFactory, &QgsChunkLoaderFactory::childrenPrepared, this, [this]
   {
@@ -86,6 +219,8 @@ QgsChunkedEntity::~QgsChunkedEntity()
   // (e.g. objects required for loading/updating have been deleted already)
   Q_ASSERT( mActiveJobs.isEmpty() );
 
+  mLog->toJson( "/tmp/dump.json" );
+
   // clean up any pending load requests
   while ( !mChunkLoaderQueue->isEmpty() )
   {
@@ -125,6 +260,10 @@ void QgsChunkedEntity::handleSceneUpdate( const SceneContext &sceneContext )
   if ( !mIsValid )
     return;
 
+  mLog->records.append( EntityUpdateRecord() );
+  mLog->currentRecord().timeUpdateStart = QTime::currentTime();
+  mLog->currentRecord().sceneContext = sceneContext;
+
   // Let's start the update by removing from loader queue chunks that
   // would get frustum culled if loaded (outside of the current view
   // of the camera). Removing them keeps the loading queue shorter,
@@ -198,6 +337,15 @@ void QgsChunkedEntity::handleSceneUpdate( const SceneContext &sceneContext )
     mBboxesEntity->setBoxes( bboxes );
   }
 
+  for ( QgsChunkNode *n : std::as_const( mActiveNodes ) )
+    mLog->currentRecord().activeNodes.append( n->tileId() );
+  _dumpList( mChunkLoaderQueue, mLog->currentRecord().loadingQueueNode );
+  _dumpList( mReplacementQueue, mLog->currentRecord().replacementQueueNodes );
+  for ( QgsChunkQueueJob *j : std::as_const( mActiveJobs ) )
+    mLog->currentRecord().activeJobsNodes.append( _dumpChunkInfo( j->chunk() ) );
+
+  mLog->currentRecord().timeUpdateFinish = QTime::currentTime();
+
   // start a job from queue if there is anything waiting
   startJobs();
 
diff --git a/src/3d/chunks/qgschunkedentity_p.h b/src/3d/chunks/qgschunkedentity_p.h
index f7bbc5f23cd..f6f2b78d6cd 100644
--- a/src/3d/chunks/qgschunkedentity_p.h
+++ b/src/3d/chunks/qgschunkedentity_p.h
@@ -47,6 +47,8 @@ namespace QgsRayCastingUtils
   struct RayHit;
 }
 
+struct EntityLog;
+
 #include <QVector3D>
 #include <QMatrix4x4>
 
@@ -167,6 +169,8 @@ class QgsChunkedEntity : public Qgs3DMapSceneEntity
     bool mIsValid = true;
 
     int mPrimitivesBudget = std::numeric_limits<int>::max();
+
+    std::unique_ptr<EntityLog> mLog;
 };
 
 /// @endcond
